### PR TITLE
customized data reduction for new data

### PR DESCRIPTION
**Pull request by marijana777**
Added NUV and FUV customized data reduction for the newest Calcos version (at that time)

### CODE REVIEW

File: cosredux/science.py [hunk 1]
```diff
@@ -248,4 +255,344 @@ def combinespectfiles(spfile_a, spfile_b, file_ab):
     #print("Wrote {:s}".format(file_ab))
 
 
+
+
+
+
+
+def flxwave(fname1, fname2s, xlim=(1900, 1950), ylim=None, norm=True, nseg=0):
```
Comment (profxj) [inline cosredux/science.py]: Am thinking you should make a new module, e.g. `coadding.py`
and put these methods in there.  Do so if you agree.

File: cosredux/science.py [hunk 2]
```diff
@@ -248,4 +255,344 @@ def combinespectfiles(spfile_a, spfile_b, file_ab):
...
+    fldpth : str
+       path to the files
+    det : 'FUV' or 'NUV'
+      detector
+    minsn : float
+      minimum S/N per resolution element, default = 1
+
+    Returns
+    -------
+    spectrasn : list of XSpectrum1D objects
```
Comment (profxj) [inline cosredux/science.py]: Looks like it comes back as an array instead of a list, agreed?

File: cosredux/science.py [hunk 3]
```diff
@@ -248,4 +255,344 @@ def combinespectfiles(spfile_a, spfile_b, file_ab):
...
+        wvmins.append(spects[i].wvmin.value)
+    ii = np.argsort(wvmins)
+    spects = spects[ii]
+
+    # Create a list of spectra (outpspects) in which spectra do not overlap in wavelenght with each other
+    # Coadd spectra whose wavelengths overlap (spects1).
+    outpspects = []
+    spects1 = [spects[0]] # a temporary list in which every next spectrum overlaps in wavelenght with the previous one
+    wvmax1 = spects[0].wvmax.value # max wavelenght in the previous spectrum
+    for i in range(len(spects) - 1):
```
Comment (profxj) [inline cosredux/science.py]: For future reference, you could have just done:

`for sp in spects[1:]:`

File: cosredux/science.py [hunk 4]
```diff
@@ -248,4 +255,344 @@ def combinespectfiles(spfile_a, spfile_b, file_ab):
...
+        pdb.set_trace()
+
+    medsn = np.median(sn)
+
+    return medsn
+
+
+
+
+def smoothsp(spects, det, snmin):
```
Comment (profxj) [inline cosredux/science.py]: A rather strange module, but ok..

File: cosredux/trace.py [hunk 5]
```diff
@@ -303,3 +376,605 @@ def traceshist(file_tr,traces_n,plottype='all',ymin=300,ymax=700, offs1ob = 0.,
...
+    hd = hdu[0].header
+    hdu.close()
+    cenwave = hd['CENWAVE']
+    opt_elem = hd['OPT_ELEM']
+
+    if hd['DETECTOR'] == 'NUV':
+        segments = ['NUVA', 'NUVB', 'NUVC']
+        LP2_1dx_file = fldpth + hd['XTRACTAB'][5:-5] + '_copy1.fits'
+        apertures = ['PSA', 'WCA']
+    if hd['DETECTOR'] == 'FUV':
```
Comment (profxj) [inline cosredux/trace.py]: probably best to use `elif` here and then `else:  raise error`

File: cosredux/trace.py [hunk 6]
```diff
@@ -303,3 +376,605 @@ def traceshist(file_tr,traces_n,plottype='all',ymin=300,ymax=700, offs1ob = 0.,
...
+    for i in range(len(slopes)):
+        if xmax is None:
+            xmax = np.max(xfull)
+        k = np.where((abs(yfull - slopes[i] * xfull - ypeaks[i]) < heights[i] / 2) &
+                     (xfull >= xmin) & (xfull <= xmax))[0]
+        yfull1 = yfull[k]
+        xfull1 = xfull[k]
+        # fit
+        if hd['DETECTOR'] == 'NUV':
+            x1, x0 = np.polyfit(xfull1, yfull1, 1)
```
Comment (profxj) [inline cosredux/trace.py]: A bit risky to fit without rejection.  One outlier
could give a bad result.

pypeit.utils has a `robust_polyfit` method to consider using.

File: cosredux/utils.py [hunk 7]
```diff
@@ -409,4 +410,144 @@ def change_dq_wgt(x1d_folder, clobber=True):
...
+    fldnames = []
+    ff = np.sort(glob.glob(pth))
+    for i in range(len(ff)):
+        if len(ff[i]) == (len(pth)+lenname-2):
+            ff1.append(ff[i])
+            fldnames.append(ff[i][len(pth)-3:-1])
+    return ff1, fldnames
+
+
+def findcorrtag(fldpth, verbose=True):
```
Comment (profxj) [inline cosredux/utils.py]: This is quite specific to our ELG program right?

If so, you should have this method in that Repository.

**Review by profxj — CHANGES_REQUESTED** [review]
Just a few changes to work on.

File: cosredux/science.py (hunk 2 above)
Comment (marijana777) [reply cosredux/science.py]: Agreed, and changed the explanation.

File: cosredux/science.py (hunk 3 above)
Comment (marijana777) [reply cosredux/science.py]: Changed.

File: cosredux/trace.py (hunk 5 above)
Comment (marijana777) [reply cosredux/trace.py]: Now using elif and else: raise error.

File: cosredux/utils.py (hunk 7 above)
Comment (marijana777) [reply cosredux/utils.py]: Now is in the appropriate repository.

File: cosredux/science.py (hunk 1 above)
Comment (marijana777) [reply cosredux/science.py]: Agreed. Made the new module.

File: cosredux/utils.py (hunk 7 above)
Comment (marijana777) [reply cosredux/utils.py]: I also did PR in this repository.

File: cosredux/trace.py (hunk 6 above)
Comment (marijana777) [reply cosredux/trace.py]: Now using pypeit.utils.robust_polyfit()

**Comment by marijana777** [comment]
@profxj all changes completed.

**Comment by profxj** [comment]
merging.

**Pull request merged**